Address some SaferCPP "Forward declared..." issues in WebKit/

geoffreygaren opened the pull request:
#### 1ae1b3ca256e32adcf9f232e9c4faacabe730c36
<pre>
Address some SaferCPP &quot;Forward declared...&quot; issues in WebKit/
<a href="https://bugs.webkit.org/show_bug.cgi?id=299641">https://bugs.webkit.org/show_bug.cgi?id=299641</a>
<a href="https://rdar.apple.com/161444598">rdar://161444598</a>

Reviewed by Ryosuke Niwa.

Adopted CheckedPtr for Style::BuilderState. In this case, the complaint about
forward declaration was a real issue with unproven lifetime for a shared object.

Moved some inline functions out of line to make all type definitions visible
without increasing #includes in headers. (These are not hot paths, and/or
virtual functions anyway.)

Added some #includes in autogenerated IPC encoders / decoders to make type
definitions visible.

* Source/JavaScriptCore/heap/AbstractSlotVisitor.h:
* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::TypeDeducingCSSValueMapper::operator unsigned short const):
(WebCore::TypeDeducingCSSValueMapper::operator int const):
(WebCore::TypeDeducingCSSValueMapper::operator unsigned const):
(WebCore::TypeDeducingCSSValueMapper::operator float const):
(WebCore::TypeDeducingCSSValueMapper::operator double const):
* Source/WebCore/css/CSSStyleProperties.cpp:
(WebCore::StyleRuleCSSStyleProperties::parentRule const):
* Source/WebCore/css/CSSStyleProperties.h:
* Source/WebCore/css/CSSToLengthConversionData.cpp:
(WebCore::CSSToLengthConversionData::protectedStyleBuilderState const):
* Source/WebCore/css/CSSToLengthConversionData.h:
(WebCore::CSSToLengthConversionData::styleBuilderState const):
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
(WebCore::CSSCalc::toCalculationValue):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate):
(WebCore::CSSCalc::evaluateWithoutFallback):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
(WebCore::CSSPropertyParserHelpers::parseFilterValueListOrNoneRaw):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp:
(WebCore::CSSPropertyParserHelpers::parseTransformRaw):
* Source/WebCore/loader/cache/CachedImageClient.cpp: Added.
(WebCore::CachedImageClient::imageFrameAvailable):
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::expectedType): Deleted.
(WebCore::CachedImageClient::imageChanged): Deleted.
(WebCore::CachedImageClient::canDestroyDecodedData const): Deleted.
(WebCore::CachedImageClient::imageFrameAvailable): Deleted.
(WebCore::CachedImageClient::imageVisibleInViewport const): Deleted.
(WebCore::CachedImageClient::didRemoveCachedImageClient): Deleted.
(WebCore::CachedImageClient::imageContentChanged): Deleted.
(WebCore::CachedImageClient::scheduleRenderingUpdateForImage): Deleted.
(WebCore::CachedImageClient::allowsAnimation const): Deleted.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
* Source/WebCore/rendering/AccessibilityRegionContext.cpp:
(WebCore::AccessibilityRegionContext::takeBounds):
* Source/WebCore/rendering/AccessibilityRegionContext.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::Builder):
(WebCore::Style::Builder::applyTopPriorityProperties):
(WebCore::Style::Builder::applyHighPriorityProperties):
(WebCore::Style::Builder::applyNonHighPriorityProperties):
(WebCore::Style::Builder::adjustAfterApplying):
(WebCore::Style::Builder::applyPropertiesImpl):
(WebCore::Style::Builder::applyCustomProperties):
(WebCore::Style::Builder::applyCustomProperty):
(WebCore::Style::Builder::applyCustomPropertyImpl):
(WebCore::Style::Builder::applyCascadeProperty):
(WebCore::Style::Builder::applyRollbackCascadeProperty):
(WebCore::Style::Builder::applyRollbackCascadeCustomProperty):
(WebCore::Style::Builder::applyProperty):
(WebCore::Style::Builder::resolveVariableReferences):
(WebCore::Style::Builder::resolveCustomPropertyForContainerQueries):
(WebCore::Style::Builder::resolveCustomPropertyValue):
(WebCore::Style::Builder::applyPageSizeDescriptor):
(WebCore::Style::Builder::ensureRollbackCascadeForRevert):
(WebCore::Style::Builder::ensureRollbackCascadeForRevertLayer):
* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPositionTryFallbacks):
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::create):
* Source/WebCore/style/StyleCustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::parseInitialValue):
* Source/WebCore/style/values/color/StyleDynamicRangeLimitMix.cpp:
(WebCore::Style::ToStyle&lt;CSS::DynamicRangeLimitMixFunction&gt;::operator):
* Source/WebCore/style/values/easing/StyleLinearEasingFunction.cpp:
(WebCore::Style::createTimingFunctionWithResolver):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/301075@main">https://commits.webkit.org/301075@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03b24a92368afd63e6e45ca4bae461009e95c05c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124867 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44538 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35274 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131713 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76768 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/820e624a-372d-485d-9517-1af174db4536) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126744 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45240 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53107 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95023 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63050 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6fd8e440-04cb-410d-8834-6adba559e133) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127821 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36095 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111673 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75580 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e9242736-0062-44dc-97c8-94a8ef36d975) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35029 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29830 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75189 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116975 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105855 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30058 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134381 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123387 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51709 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39511 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103506 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52126 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107888 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103270 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48624 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26912 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48729 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19570 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51585 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57393 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/156412 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50974 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39173 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54330 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52667 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e55134155fe6e83a35fbb1f0016b62e7377ca916. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51417

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/40d17afbed059a0d3ee6af82323618f0b91378e6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51417

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/73b4b75fdf256f2880274e967e6b18203fa96fec. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51417

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a8f07797dd8d9edc8398f50c2b6c94f18c6b8895. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51417

cdumez commented:
> #### [73b4b75](https://github.com/WebKit/WebKit/commit/73b4b75fdf256f2880274e967e6b18203fa96fec)
> aAddress some SaferCPP "Forward declared..." issues in WebKit/

Typo: `aAddress`

geoffreygaren commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1387,7 +1387,8 @@ inline NameScope BuilderConverter::convertNameScope(BuilderState& builderState,
 {
-    auto convertFallback = [&](const CSSValue& fallbackValue) -> std::optional<PositionTryFallback> {
+    // FIXME: Ideally we'd be able to mark the convertFallback() lamdba as NOESCAPE to avoid having to suppress.
+    SUPPRESS_UNCOUNTED_LAMBDA_CAPTURE auto convertFallback = [&](const CSSValue& fallbackValue) -> std::optional<PositionTryFallback> {
```
Weird! I'm not sure why I saw something different on the bot. I'll revert for now.

geoffreygaren replied:
I'll add back the suppress for now.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8907cb2d9beb836a82a56bafc15215d3dda56102. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51417

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b83730b3366c73e2453a772bb9c3ff704ded0284. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51417

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/62a368807c9d91008c8f76a32a20bcd73e066a8e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51417

geoffreygaren committed:
Address some SaferCPP "Forward declared..." issues in WebKit/
https://bugs.webkit.org/show_bug.cgi?id=299641
rdar://161444598

Reviewed by Ryosuke Niwa.

Adopted CheckedPtr for Style::BuilderState. In this case, the complaint about
forward declaration was a real issue with unproven lifetime for a shared object.

Moved some inline functions out of line to make all type definitions visible
without increasing #includes in headers. (These are not hot paths, and/or
virtual functions anyway.)

Added some #includes in autogenerated IPC encoders / decoders to make type
definitions visible.

* Source/JavaScriptCore/heap/AbstractSlotVisitor.h:
* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::TypeDeducingCSSValueMapper::operator unsigned short const):
(WebCore::TypeDeducingCSSValueMapper::operator int const):
(WebCore::TypeDeducingCSSValueMapper::operator unsigned const):
(WebCore::TypeDeducingCSSValueMapper::operator float const):
(WebCore::TypeDeducingCSSValueMapper::operator double const):
* Source/WebCore/css/CSSStyleProperties.cpp:
(WebCore::StyleRuleCSSStyleProperties::parentRule const):
* Source/WebCore/css/CSSStyleProperties.h:
* Source/WebCore/css/CSSToLengthConversionData.cpp:
(WebCore::CSSToLengthConversionData::protectedStyleBuilderState const):
* Source/WebCore/css/CSSToLengthConversionData.h:
(WebCore::CSSToLengthConversionData::styleBuilderState const):
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
(WebCore::CSSCalc::toCalculationValue):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate):
(WebCore::CSSCalc::evaluateWithoutFallback):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
(WebCore::CSSPropertyParserHelpers::parseFilterValueListOrNoneRaw):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp:
(WebCore::CSSPropertyParserHelpers::parseTransformRaw):
* Source/WebCore/loader/cache/CachedImageClient.cpp: Added.
(WebCore::CachedImageClient::imageFrameAvailable):
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::expectedType): Deleted.
(WebCore::CachedImageClient::imageChanged): Deleted.
(WebCore::CachedImageClient::canDestroyDecodedData const): Deleted.
(WebCore::CachedImageClient::imageFrameAvailable): Deleted.
(WebCore::CachedImageClient::imageVisibleInViewport const): Deleted.
(WebCore::CachedImageClient::didRemoveCachedImageClient): Deleted.
(WebCore::CachedImageClient::imageContentChanged): Deleted.
(WebCore::CachedImageClient::scheduleRenderingUpdateForImage): Deleted.
(WebCore::CachedImageClient::allowsAnimation const): Deleted.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
* Source/WebCore/rendering/AccessibilityRegionContext.cpp:
(WebCore::AccessibilityRegionContext::takeBounds):
* Source/WebCore/rendering/AccessibilityRegionContext.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::Builder):
(WebCore::Style::Builder::applyTopPriorityProperties):
(WebCore::Style::Builder::applyHighPriorityProperties):
(WebCore::Style::Builder::applyNonHighPriorityProperties):
(WebCore::Style::Builder::adjustAfterApplying):
(WebCore::Style::Builder::applyPropertiesImpl):
(WebCore::Style::Builder::applyCustomProperties):
(WebCore::Style::Builder::applyCustomProperty):
(WebCore::Style::Builder::applyCustomPropertyImpl):
(WebCore::Style::Builder::applyCascadeProperty):
(WebCore::Style::Builder::applyRollbackCascadeProperty):
(WebCore::Style::Builder::applyRollbackCascadeCustomProperty):
(WebCore::Style::Builder::applyProperty):
(WebCore::Style::Builder::resolveVariableReferences):
(WebCore::Style::Builder::resolveCustomPropertyForContainerQueries):
(WebCore::Style::Builder::resolveCustomPropertyValue):
(WebCore::Style::Builder::applyPageSizeDescriptor):
(WebCore::Style::Builder::ensureRollbackCascadeForRevert):
(WebCore::Style::Builder::ensureRollbackCascadeForRevertLayer):
* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPositionTryFallbacks):
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::create):
* Source/WebCore/style/StyleCustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::parseInitialValue):
* Source/WebCore/style/values/color/StyleDynamicRangeLimitMix.cpp:
(WebCore::Style::ToStyle<CSS::DynamicRangeLimitMixFunction>::operator):
* Source/WebCore/style/values/easing/StyleLinearEasingFunction.cpp:
(WebCore::Style::createTimingFunctionWithResolver):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/301075@main
